File: EggIncognito.CodeGen/Templates/javascript/server.test.js
'use strict';
const { test } = require('node:test');
const assert = require('node:assert/strict');
const fs = require('fs');
const path = require('path');
const os = require('os');
const server = require('./server');

const EID_PROTO = 'MhJFSTAwMDAwMDAwMDAwMDAwMDE=';

test('extractEid returns correct EID', () => {
  assert.equal(server.extractEid(EID_PROTO), 'EI0000000000000001');
});

test('extractEid returns empty string for empty input', () => {
  assert.equal(server.extractEid(''), '');
});

test('loadFixture returns default fixture', () => {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'ei-test-'));
  try {
    fs.mkdirSync(path.join(dir, 'default'));
    fs.writeFileSync(path.join(dir, 'default', 'test.binpb'), Buffer.from([0x01, 0x02]));
    server.FIXTURES_PATH = dir;
    const data = server.loadFixture('test', '');
    assert.deepEqual(data, Buffer.from([0x01, 0x02]));
  } finally {
    fs.rmSync(dir, { recursive: true });
    server.FIXTURES_PATH = process.env.FIXTURES_PATH ?? 'Fixtures';
  }
});

test('loadFixture returns EID override', () => {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'ei-test-'));
  try {
    fs.mkdirSync(path.join(dir, 'default'));
    fs.mkdirSync(path.join(dir, 'eids', 'EI0000000000000001'), { recursive: true });
    fs.writeFileSync(path.join(dir, 'default', 'test.binpb'), Buffer.from([0x01]));
    fs.writeFileSync(path.join(dir, 'eids', 'EI0000000000000001', 'test.binpb'), Buffer.from([0x02]));
    server.FIXTURES_PATH = dir;
    const data = server.loadFixture('test', 'EI0000000000000001');
    assert.deepEqual(data, Buffer.from([0x02]));
  } finally {
    fs.rmSync(dir, { recursive: true });
    server.FIXTURES_PATH = process.env.FIXTURES_PATH ?? 'Fixtures';
  }
});

test('loadFixture returns empty buffer for missing fixture', () => {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'ei-test-'));
  try {
    server.FIXTURES_PATH = dir;
    const data = server.loadFixture('nonexistent', '');
    assert.deepEqual(data, Buffer.alloc(0));
  } finally {
    fs.rmSync(dir, { recursive: true });
    server.FIXTURES_PATH = process.env.FIXTURES_PATH ?? 'Fixtures';
  }
});

File: EggIncognito.CodeGen/Templates/go/server_test.go
//go:build ignore

package main

import (
	"os"
	"path/filepath"
	"testing"
)

const eidProto = "MhJFSTAwMDAwMDAwMDAwMDAwMDE="

func TestExtractEid(t *testing.T) {
	got := extractEid(eidProto)
	if got != "EI0000000000000001" {
		t.Errorf("got %q, want EI0000000000000001", got)
	}
}

func TestExtractEidEmpty(t *testing.T) {
	if extractEid("") != "" {
		t.Errorf("expected empty string for empty input")
	}
}

func TestLoadFixtureDefault(t *testing.T) {
	dir := t.TempDir()
	defDir := filepath.Join(dir, "default")
	os.MkdirAll(defDir, 0755)
	os.WriteFile(filepath.Join(defDir, "test.binpb"), []byte{0x01, 0x02}, 0644)

	orig := fixturesPath
	fixturesPath = dir
	t.Cleanup(func() { fixturesPath = orig })

	data := loadFixture("test", "")
	if len(data) != 2 || data[0] != 0x01 || data[1] != 0x02 {
		t.Errorf("unexpected fixture data: %v", data)
	}
}

func TestLoadFixtureEidOverride(t *testing.T) {
	dir := t.TempDir()
	os.MkdirAll(filepath.Join(dir, "default"), 0755)
	os.MkdirAll(filepath.Join(dir, "eids", "EI0000000000000001"), 0755)
	os.WriteFile(filepath.Join(dir, "default", "test.binpb"), []byte{0x01}, 0644)
	os.WriteFile(filepath.Join(dir, "eids", "EI0000000000000001", "test.binpb"), []byte{0x02}, 0644)

	orig := fixturesPath
	fixturesPath = dir
	t.Cleanup(func() { fixturesPath = orig })

	data := loadFixture("test", "EI0000000000000001")
	if len(data) != 1 || data[0] != 0x02 {
		t.Errorf("expected EID fixture byte 0x02, got %v", data)
	}
}

func TestLoadFixtureMissing(t *testing.T) {
	dir := t.TempDir()
	orig := fixturesPath
	fixturesPath = dir
	t.Cleanup(func() { fixturesPath = orig })

	data := loadFixture("nonexistent", "")
	if len(data) != 0 {
		t.Errorf("expected empty bytes for missing fixture, got %v", data)
	}
}
